src/parquet/file/metadata-internal.cc

```diff
@@ -0,0 +1,500 @@
+      : row_group_(row_group) {}
+  ~RowGroupMetaDataImpl() {}
+
+  int num_columns() const { return row_group_->columns.size(); }
+
+  int64_t num_rows() const { return row_group_->num_rows; }
+
+  int64_t total_byte_size() const { return row_group_->total_byte_size; }
+
+  std::unique_ptr<ColumnMetaData> GetColumnMetaData(int i) {
+    DCHECK(i < num_columns()) << "The file only has " << num_columns()
+                              << " columns, requested metadata for column: " << i;
+    return ColumnMetaData::GetColumnMetaData(
+        reinterpret_cast<const uint8_t*>(&row_group_->columns[i]));
+  }
+
+ private:
+  const format::RowGroup* row_group_;
+};
+
+std::unique_ptr<RowGroupMetaData> RowGroupMetaData::GetRowGroupMetaData(
+    const uint8_t* metadata) {
+  return std::unique_ptr<RowGroupMetaData>(new RowGroupMetaData(metadata));
+}
+
+RowGroupMetaData::RowGroupMetaData(const uint8_t* metadata)
+    : impl_{std::unique_ptr<RowGroupMetaDataImpl>(new RowGroupMetaDataImpl(
+          reinterpret_cast<const format::RowGroup*>(metadata)))} {}
+RowGroupMetaData::~RowGroupMetaData() {}
+
+int RowGroupMetaData::num_columns() const {
+  return impl_->num_columns();
+}
+
+int64_t RowGroupMetaData::num_rows() const {
+  return impl_->num_rows();
+}
+
+int64_t RowGroupMetaData::total_byte_size() const {
+  return impl_->total_byte_size();
+}
+
+std::unique_ptr<ColumnMetaData> RowGroupMetaData::GetColumnMetaData(int i) const {
+  return impl_->GetColumnMetaData(i);
+}
+
+// file metadata
+class FileMetaData::FileMetaDataImpl {
+ public:
+  explicit FileMetaDataImpl(const uint8_t* metadata)
+      : metadata_{std::unique_ptr<const format::FileMetaData>(
+            reinterpret_cast<const format::FileMetaData*>(metadata))} {
+    schema::FlatSchemaConverter converter(
+        &metadata_->schema[0], metadata_->schema.size());
+    schema_.Init(converter.Convert());
+  }
+  ~FileMetaDataImpl() {}
+  int64_t num_rows() const { return metadata_->num_rows; }
+  int num_row_groups() const { return metadata_->row_groups.size(); }
+  int32_t version() const { return metadata_->version; }
+  const std::string& created_by() const { return metadata_->created_by; }
+  int num_schema_elements() const { return metadata_->schema.size(); }
+
+  void WriteTo(OutputStream* dst) { SerializeThriftMsg(metadata_.get(), 1024, dst); }
+
+  std::unique_ptr<RowGroupMetaData> GetRowGroupMetaData(int i) {
+    DCHECK(i < num_row_groups())
+        << "The file only has " << num_row_groups()
+        << " row groups, requested metadata for row group: " << i;
+    return RowGroupMetaData::GetRowGroupMetaData(
+        reinterpret_cast<const uint8_t*>(&metadata_->row_groups[i]));
+  }
+
+  const SchemaDescriptor* schema() const { return &schema_; }
+
+ private:
+  std::unique_ptr<const format::FileMetaData> metadata_;
+  SchemaDescriptor schema_;
+};
+
+std::unique_ptr<FileMetaData> FileMetaData::GetFileMetaData(const uint8_t* metadata) {
```

Comment:
I will rename them to `Make`. I made them `unique_ptr` for the only reason that they are more efficient than `shared_ptr`.